### PR TITLE
tighten what is considered a platform extern.

### PR DESCRIPTION
**Pull request by rkirov**
Adds an extra file general_with_platform.d.ts which will contain all
expected output from the platfrom externs.

### CODE REVIEW

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 1]
```diff
@@ -70,6 +70,13 @@
   public static final Pattern GOOG_MODULE_EXTRACT =
       Pattern.compile("goog.module\\(['\"](.*)['\"]\\);");
 
+  /**
+   * Contains symbols that are part of platform externs, but not yet in lib.d.ts
+   * This list is incomplete and will grow as needed.
+   */
+  private static final Set<String> emittedPlatformSymbols = Sets.newHashSet(
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: Why a Set for this, but a switch for the files? And if a Set, why not an immutable set?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 2]
```diff
@@ -409,14 +416,39 @@ private boolean isDefaultExport(TypedVar symbol) {
...
 
-  // For platform externs we skip emitting .d.ts, to avoid collisions with lib.d.ts.
-  // TODO(rado): Platform is too ill-defined, and this often filters too much.
-  // Replace with a list of externs that *only* contain symbols in lib.d.ts.
-  private boolean isPlatformExtern(String name) {
-    name = name.replace("externs.zip//", "");
-    // mostly matching what is in https://github.com/google/closure-compiler/tree/master/externs.
-    return Pattern.matches("javascript/externs/[^/]*.js", name) || name.startsWith("es")
-        || name.startsWith("w3c") || name.startsWith("ie_") || name.startsWith("browser");
+  // For platform externs we skip emitting, to avoid collisions with lib.d.ts.
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: JavaDoc

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 3]
```diff
@@ -409,14 +416,39 @@ private boolean isDefaultExport(TypedVar symbol) {
...
+  private boolean isPlatformExtern(String filePath, String symbolName) {
+    // Some symbols are not yet provided by lib.d.ts, so we have to emit them.
+    if (emittedPlatformSymbols.contains(symbolName)) return false;
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: I think calling this `platformSymbolsMissingInTypeScript` would make more sense to me.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 4]
```diff
@@ -70,6 +70,13 @@
   public static final Pattern GOOG_MODULE_EXTRACT =
       Pattern.compile("goog.module\\(['\"](.*)['\"]\\);");
 
+  /**
+   * Contains symbols that are part of platform externs, but not yet in lib.d.ts
+   * This list is incomplete and will grow as needed.
+   */
+  private static final Set<String> emittedPlatformSymbols = Sets.newHashSet(
+      "Image");
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: Would it make sense to just include all the ones that you need right now?

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java [hunk 5]
```diff
@@ -409,14 +416,39 @@ private boolean isDefaultExport(TypedVar symbol) {
...
-    // mostly matching what is in https://github.com/google/closure-compiler/tree/master/externs.
-    return Pattern.matches("javascript/externs/[^/]*.js", name) || name.startsWith("es")
-        || name.startsWith("w3c") || name.startsWith("ie_") || name.startsWith("browser");
+  // For platform externs we skip emitting, to avoid collisions with lib.d.ts.
+  private boolean isPlatformExtern(String filePath, String symbolName) {
+    // Some symbols are not yet provided by lib.d.ts, so we have to emit them.
+    if (emittedPlatformSymbols.contains(symbolName)) return false;
+    // This matches our test setup.
+    if (filePath.startsWith("externs.zip//")) return true;
+    String fileName = Iterables.getLast(Splitter.on("/").split(filePath));
```
Comment (mprobst) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: This is called hundreds of times during one run, right? What about `String fileName = filePath.substring(filePath.lastIndexOf('/') + 1)`? Alternatively, `new File(filePath).getName()`?

**Comment by mprobst** [comment]
LGTM.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 1 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: made them all immutable sets.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 2 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 3 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 4 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: added 8 other ones, which lead to 12 total, general_with_platform.d.ts bubbled a bit.

File: src/main/java/com/google/javascript/clutz/DeclarationGenerator.java (hunk 5 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/clutz/DeclarationGenerator.java]: done

**Pull request merged**